aws/hll/aws-s3-transfer-manager/src/download.rs

```diff
@@ -34,17 +49,178 @@ impl From<GetObjectInputBuilder> for DownloadRequest {
+#[derive(Debug, Clone)]
+pub struct Downloader {
+    target_part_size_bytes: u64,
+    checksum_validation_enabled: bool,
+    concurrency: usize,
+    client: aws_sdk_s3::client::Client,
+}
+
+impl Downloader {
+    /// Create a new [Builder]
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+
+    /// Download a single object from S3.
+    ///
+    /// A single logical request may be split into many concurrent ranged `GetObject` requests
+    /// to improve throughput.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::error::Error;
+    /// use aws_sdk_s3::operation::get_object::builders::GetObjectInputBuilder;
+    /// use aws_s3_transfer_manager::download::{Downloader, DownloadRequest};
+    ///
+    /// async fn get_object(client: Downloader) -> Result<(), Box<dyn Error>> {
+    ///     let request = GetObjectInputBuilder::default()
+    ///         .bucket("my-bucket")
+    ///         .key("my-key")
+    ///         .into();
+    ///
+    ///     let handle = client.download(request).await?;
+    ///     // process data off handle...
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn download(&self, req: DownloadRequest) -> Result<DownloadHandle, TransferError> {
+        // if there is a part number then just send the default request
+        if req.input.get_part_number().is_some() {
+            todo!("single part download not implemented")
+        }
+
+        let ctx = DownloadContext {
+            client: self.client.clone(),
+            target_part_size: self.target_part_size_bytes,
+        };
+
+        // make initial discovery about the object size, metadata, possibly first chunk
+        let mut discovery = discover_obj(&ctx, &req).await?;
+        let (comp_tx, comp_rx) = mpsc::channel(self.concurrency);
+        let start_seq = handle_discovery_chunk(&mut discovery, &comp_tx).await;
+
+        // spawn all work into the same JoinSet such that when the set is dropped all tasks are cancelled.
+        let mut tasks = JoinSet::new();
+
+        if !discovery.remaining.is_empty() {
+            // start assigning work
+            let (work_tx, work_rx) = async_channel::bounded(self.concurrency);
+            let input = req.input.clone();
+            let part_size = self.target_part_size_bytes;
+            let rem = discovery.remaining.clone();
+
+            tasks.spawn(distribute_work(rem, input, part_size, start_seq, work_tx));
+
+            for i in 0..self.concurrency {
+                let worker = chunk_downloader(ctx.clone(), work_rx.clone(), comp_tx.clone())
+                    .instrument(tracing::debug_span!("chunk-downloader", worker = i));
+                tasks.spawn(worker);
+            }
+        }
+
+        // drop our reference to completion sender such that when all workers drop theirs it's closed
+        drop(comp_tx);
+
+        let handle = DownloadHandle {
+            object_meta: discovery.meta,
```

Comment:
TODO: this needs fixed, the initial object discovery is for a range/first-part which will not have the right metadata w.r.t content-length, etc. I'm not sure what we want to do here or if there are any other metadata fields that would vary when doing a "full" get object vs ranged GET(s). Presumably CRT team already knows answer to this.